### PR TITLE
Instead of highly specialized to libFuzzer, make function-based fuzzing interface more generic

### PR DESCRIPTION
**Pull request by agroce**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -23,6 +23,10 @@ if (NOT DEFINED BUILD_NOSTATIC AND DEFINED ENV{BUILD_NOSTATIC})
   set(BUILD_NOSTATIC "$ENV{BUILD_NOSTATIC}") 
 endif()
 
+if (NOT DEFINED BUILD_NO32 AND DEFINED ENV{BUILD_NO32}) 
+  set(BUILD_NO32 "$ENV{BUILD_NOSTATIC}")
```
Comment (pgoodman) [inline CMakeLists.txt]: `set(BUILD_NO32 "$ENV{BUILD_NO32}")`?

File: CMakeLists.txt [hunk 2]
```diff
@@ -67,49 +71,83 @@ if (NOT WIN32)
   set(CMAKE_CXX_STANDARD 11)
 endif()
 
-add_library(${PROJECT_NAME} STATIC
-  src/lib/DeepState.c
-  src/lib/Log.c
-  src/lib/Option.c
-  src/lib/Stream.c
-)
+if (NOT BUILD_NOSTATIC)
```
Comment (pgoodman) [inline CMakeLists.txt]: This double negative here is a bit unusual. 

I think it's common in CMake to have `BUILD_STATIC` and `BUILD_SHARED` options, along with `BUILD_32BIT` (which might produce artifacts for both the shared and static case?). Like, maybe we want:

```cmake
option(DEEPSTATE_BUILD_SHARED "Enable build of shared libraries" ON)
option(DEEPSTATE_BUILD_STATIC "Enable build of static libraries" ON)
```

Then we could perhaps have different target names, for exmaple: https://github.com/facebook/rocksdb/blob/master/CMakeLists.txt#L724-L726 (ignoring `${ARTIFACT_SUFFIX}`) .

The options can be set at the command line invocation of cmake, with `-DDEEPSTATE_ENABLE_SHARED=ON` or `..=OFF`. Similarly, it enables another project to include our code and configure it in CMake.

@alessandrogario might have some good tips on this.

**Comment by agroce** [comment]
Good idea, in fact the whole CMakeLists.txt could probably use a refactor by someone with less cargo-culty cmake skills than I have... :)

File: CMakeLists.txt (hunk 2 above)
Comment (agroce) [reply CMakeLists.txt]: (Agreed)

File: CMakeLists.txt (hunk 1 above)
Comment (pgoodman) [reply CMakeLists.txt]: @agroce is this intentional?

Comment (agroce) [reply CMakeLists.txt]: Oops no.  I will have to fix this whole mess up sometime soon

**Comment by agroce** [comment]
Sort of outdated, I doubt I'll finish this off and work on CMake

**Pull request closed without merging**